### PR TITLE
Design document for the QUIC-TLS integration

### PR DESCRIPTION
**Pull request by mattcaswell**
This provides the design document for how we will integrate the TLS handshake into QUIC.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/designs/quic-design/quic-tls.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+A QUIC-TLS handshake is managed by a QUIC_TLS object. This object provides
+3 core functions to the rest of the QUIC implementation:
+
+```c
+QUIC_TLS *ossl_quic_tls_new(SSL *s, const QUIC_TLS_ARGS *args);
+```
+
+The `ossl_quic_tls_new` function instantiates a new `QUIC_TLS` object associated
+with the QUIC Connection held in `s` and initialises it with a set of callbacks
```
Comment (hlandau) [inline doc/designs/quic-design/quic-tls.md:20]: So `s` is a `QUIC_CONNECTION`-based SSL object? Is there some reason `QUIC_TLS` needs to access the entire `QUIC_CONNECTION` &mdash; wouldn't a `SSL_CONNECTION` suffice?

May as well pass it in the args struct for consistency. So basically what I see here is the CSM creating a `SSL_CONNECTION` and then a `QUIC_TLS` object which receives a pointer to the `SSL_CONNECTION`.

**Review by hlandau — CHANGES_REQUESTED** [review]
This looks great. I'm pleasantly surprised by how closely this resembles my existing prototype, I was expecting an interface based on direct use of an SSL object. Having this separate like this certainly keeps the CSM simpler.

File: doc/designs/quic-design/quic-tls.md (hunk 1 above)
Comment (mattcaswell) [reply doc/designs/quic-design/quic-tls.md:20]: > So s is a QUIC_CONNECTION-based SSL object?

No. `s` here is a pure `SSL_CONNECTION` based SSL object.

To try and explain this, let me refer you to `ossl_quic_new` which creates the "user" visible SSL object:

https://github.com/openssl/openssl/blob/1a298b00112e50718acc1fdd40b9bce482112cbf/ssl/quic/quic_impl.c#L14-L41

In the above code `qc->tls` is the `SSL_CONNECTION` based SSL object.

So, the "user" visible SSL object has a `QUIC_CONNECTION` in it. Inside that there is an "inner" SSL object that has an `SSL_CONNECTION` in it. The above `ossl_quic_new` code will actually need some tweaking to ensure the "inner" SSL object has the right SSL_METHOD associated with it (I have those tweaks on my local machine).

We need the "inner" SSL object so we can call public API calls. We need to be able to call functions such as `SSL_connect()` and it end up calling the TLS version of `SSL_connect` not the QUIC version.

I will update the document to try and explain this.

Comment (mattcaswell) [reply doc/designs/quic-design/quic-tls.md]: I've updated the text to clarify this. Please take another look.

**Review by hlandau — APPROVED** [review]
Perfect!

Advance warning, I've added an `int is_server;` argument to the dummy handshake layer args, so if you want to do that here, go ahead.

File: doc/designs/quic-design/quic-tls.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+has successfully completed this will be communicated via `handshake_complete_cb`.
+
+In the event that an error occurs a normal TLS handshake would send a TLS alert
+record. QUIC handles this differently and so the QUIC_TLS object will intercept
+attempts to send an alert and will communicate this via the `alert_cb` callback.
+
+QUIC requires the use of a TLS extension in order to send and receive "transport
+parameters". These transport parameters are opaque to the `QUIC_TLS` object. It
+does not need to use them directly but instead simply includes them in an
+extenstion to be sent in the ClientHello and receives them back from the peer in
```
Comment (t8m) [inline doc/designs/quic-design/quic-tls.md:136]: Typo: extension

File: doc/designs/quic-design/quic-tls.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                                      void *rlarg);
+```
+
+The internal function `ssl_select_next_record_layer` which is used in the TLS
+implementation to work out which record method should be used next is modified
+to first check whether a custom record method has been specified and always use
+that one if so.
+
+The TLS record layer code is further modified to provide the following
+capabilities which are needed in order support QUIC.
```
Comment (t8m) [inline doc/designs/quic-design/quic-tls.md:172]: `in order to support`

**Review by t8m — CHANGES_REQUESTED** [review]
Just two nits

**Comment by mattcaswell** [comment]
Fixup pushed addressing the feedback.

@t8m - please take another look
@hlandau - please reconfirm

**Review by hlandau — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**